### DISCUSSION TITLE
Infinite redirection loop with <Auth /> wrapper when using custom sign-in page

### DISCUSSION THREAD

**Discussion by mark-voltacharging** [nextauthjs/next-auth#4380] [Help]
This pattern seems to give me an infinite loop of redirection if I have a custom sign-in page, has anyone else experienced this?

`pages: {
    signIn: '/auth/sign-in'
  }`

_Originally posted by @mark-voltacharging in https://github.com/nextauthjs/next-auth/issues/1210#issuecomment-1079283299_

**Comment by mark-voltacharging** [comment]
For more immediate context, this refers to an issue with this pattern

```
// In _app.js

<Auth>
      <Component {...pageProps} />
</Auth>

// The Auth Wrapper
const Auth: FC = ({ children }) => {
  const { data: session, status } = useSession({ required: true })
  const isUser = session?.user

  if (isUser) {
    return (children as React.ReactElement)
  }

  // Session is being fetched, or no user.
  // If no user, useEffect() will redirect.
  return null
}
```

I like the concept of this pattern because for my particular app, I'd like the whole thing to be restricted in an elegant way. However I also need a custom SignIn page, and this is where the problem comes in. If I adopt the custom SignIn page with
`pages: { signIn: '/auth/sign-in' }`, I get an infinite loop of redirection. This is probably because my `<SignIn />` component is a child of `<Auth />`. My browser URL quickly turns into

```
http://localhost:4200/api/auth/signin?error=SessionRequired&callbackUrl=http%3A%2F%2Flocalhost%3A4200%2Fauth%2Fsign-in%3FcallbackUrl%3Dhttp%3A%2F%2Flocalhost%3A4200%2Fauth%2Fsign-in%3FcallbackUrl%3Dhttp%3A%2F%2Flocalhost%3A4200%2Fauth%2Fsign-in%3FcallbackUrl%3Dhttp%3A%2F%2Flocalhost%3A4200%2Fauth%2Fsign-in%3FcallbackUrl%3Dhttp%3A%2F%2Flocalhost%3A4200%2Fauth%2Fsign-... and so on
```

**Reply by islamneddar** [reply]
as <SignIn/> is a child of your <Auth/> and you have  useSession({ required: true }) 
so when it arrives on SignIn too it will check if you are authenticated meanwhile redirecting because the required is true 

to solve it you have to create 2 sub trees one for your auth pages with a useSession(required: false) and the other for the protected pages with useSession(required: true)

**Comment by RadasLove** [comment]
I have exactly same problem :-/ i thought i will use 

```
some page with auth required

IndexPage.auth = {
  unauthorized: '/auth/login, // redirect to this url
};
```

From https://github.com/nextauthjs/next-auth/issues/1210  or https://next-auth.js.org/getting-started/client

but redirect dont work :-/ because i dont want have callback error parameter in URL

```
_app.tsx

function Auth({ children }) {
  // if `{ required: true }` is supplied, `status` can only be "loading" or "authenticated"
  const { status } = useSession({ required: true });

  if (status === 'loading') {
    return <div>Loading...</div>;
  }

  return children;
}
```
```
[..nextuath].js

  pages: {
    signIn: '/auth/login',
  },
```

**Comment by nahidtopalovic** [comment]
This should solve your issue
 `const page = getLayout(<Component {...pageProps} />)`
`{Component.auth ? <Auth auth={Component.auth}>{page}</Auth> : page}`

**Reply by mark-voltacharging** [reply]
hmm maybe we would be better to have a pattern that whitelists ''auth'' related components like SignIn? This way all <Components /> are blocked by default, and a few of them can be whitelisted to avoid the wrapper. Something like this

```
// _app.ts
...
{ Component.isAuth ? <Component {...pageProps} /> : <Auth><Component {...pageProps} /></Auth> }

// In SignIn.tsx
...
SignIn.isAuth = true
```

**Comment by redx1t** [comment]
It seems to be a next js framework issue. Set your framework version to 12.2.2 in package.json and npm update. 

```
{
"next": "12.2.2",
}
```

**Comment by Manitej66** [comment]
any fix for this issue?

**Comment by TomFCarrion** [comment]
I'm having exact same problem on my project with `"next": "^12.3.4", "next-auth": "^4.22.0"`, has anyone found a fix?
